src/dune_engine/target_promotion.ml

```diff
@@ -168,6 +191,10 @@ let promote ~dir ~files_to_promote ~promote ~promote_source =
           [ Pp.textf "Promoting %S to %S" (Path.Build.to_string src)
               (Path.Source.to_string dst)
           ];
+        (* We know that [dir] exists but if [dst] comes from a directory target,
+           we may need to create some additional subdirectories. *)
+        let dst_path = Path.source dst in
+        create_parent_directory_if_needed ~dst_path;
```

Comment:
If we produce one directory with a 1000 target in it, we are going to end up calling `mkdir_p` 1000 times.

Reply:
Indeed. This is now fixed: we now create directories by iterating `dirs` of the given `Targets.Produced.t`.